File: qt-widgets/hello.cpp
// Bench hello app - Qt6 Widgets variant: minimal static window, one
// "Hello" label + one button. Isolates the startup floor and baseline
// memory.
// First-presented-frame proxy: first paintEvent hit on the top-level
// widget (raster QWidget windows do not emit QWindow::frameSwapped) -
// same proxy as the other Qt apps.

#include <QtWidgets>

#include <functional>

#include "bench_common.h"

class HelloWindow : public QWidget {
public:
    std::function<void()> onFirstPaint;

protected:
    void paintEvent(QPaintEvent *e) override {
        QWidget::paintEvent(e);
        if (!m_painted) {
            m_painted = true;
            if (onFirstPaint) onFirstPaint();
        }
    }

private:
    bool m_painted = false;
};

int main(int argc, char **argv) {
    const Clock::time_point t0 = Clock::now();

    QApplication app(argc, argv);
    const Mode mode = parseMode(argc, argv);

    HelloWindow w;
    w.setWindowTitle(QStringLiteral("Bench"));
    w.resize(800, 600);

    auto *layout = new QVBoxLayout(&w);
    layout->addStretch();
    auto *label = new QLabel(QStringLiteral("Hello"));
    QFont f = label->font();
    f.setBold(true);
    f.setPixelSize(18);
    label->setFont(f);
    label->setAlignment(Qt::AlignHCenter);
    auto *btn = new QPushButton(QStringLiteral("Press"));
    layout->addWidget(label, 0, Qt::AlignHCenter);
    layout->addSpacing(12);
    layout->addWidget(btn, 0, Qt::AlignHCenter);
    layout->addStretch();

    w.onFirstPaint = [t0, mode] {
        const Clock::time_point now = Clock::now();
        printFirstFrame();
        if (mode == Mode::Startup) {
            printStartupAndExit(t0, now);
        }
    };

    w.show();
    return app.exec();
}

File: egui/src/bin/hello.rs
//! Bench hello app - egui/eframe variant: minimal static window, one
//! "Hello" label + one button. Isolates the startup floor and baseline
//! memory. Same first-frame proxy as the other egui apps (end of the
//! first update pass).

use std::time::Instant;

use bench_egui::{parse_mode, print_first_frame, print_startup_and_exit, Mode};
use eframe::egui;

struct Hello {
    t0: Instant,
    mode: Mode,
    first_frame: Option<Instant>,
}

impl eframe::App for Hello {
    fn ui(&mut self, root: &mut egui::Ui, _frame: &mut eframe::Frame) {
        let now = Instant::now();
        egui::CentralPanel::default().show(root, |ui| {
            ui.vertical_centered(|ui| {
                ui.add_space(ui.available_height() / 2.0 - 30.0);
                ui.label(egui::RichText::new("Hello").strong().size(18.0));
                let _ = ui.button("Press");
            });
        });
        if self.first_frame.is_none() {
            self.first_frame = Some(now);
            print_first_frame();
            if self.mode == Mode::Startup {
                print_startup_and_exit(self.t0, now);
            }
        }
    }
}

fn main() -> eframe::Result {
    let t0 = Instant::now();
    let mode = parse_mode();
    let options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default()
            .with_inner_size([800.0, 600.0])
            .with_title("Bench"),
        ..Default::default()
    };
    eframe::run_native(
        "Bench",
        options,
        Box::new(move |_cc| {
            Ok(Box::new(Hello {
                t0,
                mode,
                first_frame: None,
            }))
        }),
    )
}
